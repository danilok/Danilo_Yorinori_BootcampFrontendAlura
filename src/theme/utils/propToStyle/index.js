import breakpointsMedia from '../breakpointsMedia';
import { breakpoints as breakpointsTheme } from '../..';

export default function propToStyle(propName) {
  return (props) => {
    const propValue = props[propName];

    if (typeof propValue === 'string' || typeof propValue === 'number') {
      return {
        [propName]: propValue,
      };
    }

    if (typeof propValue === 'object') {
      const breakpoints = {};

      Object.keys(breakpointsTheme).forEach((breakpoint) => {
        if (propValue[breakpoint]) breakpoints[breakpoint] = { [propName]: propValue[breakpoint] };
      });

      return breakpointsMedia(breakpoints);
    }

    return '';
  };
}
