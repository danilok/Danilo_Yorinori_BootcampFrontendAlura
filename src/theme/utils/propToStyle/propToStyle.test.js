import propToStyle from '.';

describe('propToStyle()', () => {
  describe('when receives no argument', () => {
    test('returns empty string', () => {
      const propToStyleResult = propToStyle('textAlign');
      // <Text />
      const componentProps = { };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toEqual('');
    });
  });

  describe('when receives an simple argument', () => {
    test('and it is a string', () => {
      const propToStyleResult = propToStyle('textAlign');
      // <Text align="center" />
      const componentProps = { textAlign: 'center' };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toEqual({ textAlign: 'center' });
    });
    test('and it is a number', () => {
      const propToStyleResult = propToStyle('flex');
      // <Text flex={1} />
      const componentProps = { flex: 1 };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toEqual({ flex: 1 });
    });
  });

  describe('when receives an argument with breakpoints', () => {
    test('renders only one breakpoint resolution', () => {
      const propToStyleResult = propToStyle('textAlign');
      // <Text align="{ xs: 'center', }" />
      const componentProps = { textAlign: { xs: 'center' } };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toMatchSnapshot();
    });
    test('renders two or more breakpoints resolutions', () => {
      const propToStyleResult = propToStyle('textAlign');
      // <Text align="{ xs: 'center', md: 'right' }" />
      const componentProps = { textAlign: { xs: 'center', md: 'right' } };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toMatchSnapshot();
    });
    test('renders all breakpoints resolutions', () => {
      const propToStyleResult = propToStyle('textAlign');
      // <Text align="{ xs: 'center', md: 'right' }" />
      const componentProps = { textAlign: { xs: 'center', md: 'right', sm: 'center', lg: 'right', xl: 'left' } };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toMatchSnapshot();
    });
  });

  describe('when receives arguments', () => {
    describe('but none for tested property', () => {
      test('renders an empty string', () => {
        const propToStyleResult = propToStyle('textAlign');
        // <Text width="50px" padding="10px" />
        const componentProps = { width: '50px', padding: '10px' };
        const styleResult = propToStyleResult(componentProps);
        expect(styleResult).toEqual('');
      });
    });
  });
});
