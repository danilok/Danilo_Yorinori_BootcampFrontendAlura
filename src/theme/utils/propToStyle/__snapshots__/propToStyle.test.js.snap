// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propToStyle() when receives an argument with breakpoints renders all breakpoints resolutions 1`] = `
Array [
  Array [
    "@media screen and (min-width:",
    "0",
    "px){",
    "text-align: center;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "480",
    "px){",
    "text-align: center;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "768",
    "px){",
    "text-align: right;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "992",
    "px){",
    "text-align: right;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "1200",
    "px){",
    "text-align: left;",
    "}",
  ],
]
`;

exports[`propToStyle() when receives an argument with breakpoints renders only one breakpoint resolution 1`] = `
Array [
  Array [
    "@media screen and (min-width:",
    "0",
    "px){",
    "text-align: center;",
    "}",
  ],
]
`;

exports[`propToStyle() when receives an argument with breakpoints renders two or more breakpoints resolutions 1`] = `
Array [
  Array [
    "@media screen and (min-width:",
    "0",
    "px){",
    "text-align: center;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "768",
    "px){",
    "text-align: right;",
    "}",
  ],
]
`;
